src/main/java/org/prebid/cache/handlers/GetCacheHandler.java

```diff
@@ -51,32 +78,108 @@ public GetCacheHandler(final ReactiveRepository<PayloadWrapper, String> reposito
+                .flatMap(GetCacheHandler::fromClientResponse)
+                .doOnError(error -> {
+                    metricsRecorder.getProxyFailure().mark();
+                    log.info("Failed to send request: '{}', cause: '{}'",
+                            ExceptionUtils.getMessage(error), ExceptionUtils.getMessage(error));
+
+                });
+    }
+
+    private void validateServerResponse(final ClientResponse clientResponse,
+                                        final SynchronousSink<ClientResponse> sink) {
+        if (HttpStatus.OK.equals(clientResponse.statusCode())) {
+            metricsRecorder.getProxySuccess().mark();
+            sink.next(clientResponse);
+            return;
+        }
+        sink.error(new BadRequestException(
+                String.format("Unexpected status code: '%s', in response from remote server",
+                        clientResponse.statusCode())));
+
+    }
+
+    private static Mono<ServerResponse> fromClientResponse(final ClientResponse clientResponse) {
+        return ServerResponse.status(clientResponse.statusCode())
+                .headers(headerConsumer -> clientResponse.headers().asHttpHeaders().forEach(headerConsumer::addAll))
+                .body(clientResponse.bodyToMono(String.class), String.class);
+    }
 
+    private Mono<ServerResponse> processRequest(final ServerRequest request, final String keyIdParam) {
+        val normalizedId = String.format("%s%s", config.getPrefix(), keyIdParam);
+        return repository.findById(normalizedId)
+                .transform(CircuitBreakerOperator.of(circuitBreaker))
+                .timeout(Duration.ofMillis(config.getTimeoutMs()))
+                .subscribeOn(Schedulers.parallel())
+                .transform(this::validateErrorResult)
+                .flatMap(wrapper -> {
+                    if (wrapper.getPayload().getType().equals(PayloadType.JSON.toString())) {
+                        metricsRecorder.markMeterForTag(this.metricTagPrefix,
+                                MetricsRecorder.MeasurementTag.JSON);
+                        return builder.createResponseMono(request,
+                                MediaType.APPLICATION_JSON_UTF8,
+                                wrapper);
+                    } else if (wrapper.getPayload()
+                            .getType()
+                            .equals(PayloadType.XML.toString())) {
+                        metricsRecorder.markMeterForTag(this.metricTagPrefix,
+                                MetricsRecorder.MeasurementTag.XML);
+                        return builder.createResponseMono(request,
+                                MediaType.APPLICATION_XML,
+                                wrapper);
+                    } else {
+                        // unhandled media type
+                        return Mono.error(new UnsupportedMediaTypeException(UNSUPPORTED_MEDIATYPE));
+                    }
+                })
```

Comment:
Please, extract to separate method. Smth like `wrapperToServerResponse`, or `fromWrapperResponse`